addons/audio/alsa.c

```diff
@@ -875,6 +878,94 @@ static void alsa_deallocate_recorder(ALLEGRO_AUDIO_RECORDER *r)
    snd_pcm_close(data->capture_handle);
 }
 
+static void _device_list_dtor(void* value, void* userdata)
+{
+   (void)userdata;
+
+   ALLEGRO_AUDIO_DEVICE* device = (ALLEGRO_AUDIO_DEVICE*)value;
+   al_free(device->name);
+   al_free(device->identifier);
+}
+
+static _AL_LIST* alsa_get_devices(void)
+{
+   if (!device_list) {
+      device_list = _al_list_create();
+
+      int rcard = -1;
+      while(snd_card_next(&rcard) == 0) {
+         if (rcard < 0) return device_list;
+
+         snd_ctl_t *handle;
+         char str[64];
+
+         sprintf(str, "hw:%i", rcard);
+         if (snd_ctl_open(&handle, str, 0) < 0) {
+            return device_list;
+         }
+
+         snd_ctl_card_info_t *card_info;
+         snd_ctl_card_info_alloca(&card_info);
+         if (snd_ctl_card_info(handle, card_info) < 0) {
+            continue;
+         }
+
+         snd_ctl_card_info_get_name(card_info);
+         snd_ctl_card_info_get_id(card_info);
+
+         int dev_num = -1;
+
+         while(1) {
+            if (snd_ctl_pcm_next_device(handle, &dev_num) < 0) {
+               break;
+            }
+
+            if (dev_num < 0) break;
+
+            void **hints;
+            int success = snd_device_name_hint(dev_num, "pcm", &hints);
+            if (success < 0) continue;
+            char**n = (char**)hints;
+               
+            char* identifier = 0;
+            char* name = 0;
+            while(*n != NULL)
+            {
+               identifier = snd_device_name_get_hint(*n, "NAME");
+               name = snd_device_name_get_hint(*n, "DESC");
```

Comment:
Why is DESC treated as a name, rather than NAME?

Reply:
For ALSA,  the 'name' is a (presumably unique) system id, and 'desc' is a friendly name, for example I have name=`alsa_output.pci-0000_00_05.0.analog-stereo` and desc=`Built-in Audio Analog Stereo`
So I think the friendly name is the one that the user should see.
I haven't checked other platforms though,